enterprise/coderd/prebuilds/claim.go

```diff
@@ -0,0 +1,64 @@
+package prebuilds
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+
+	"github.com/coder/coder/v2/coderd/prebuilds"
+
+	"github.com/google/uuid"
+	"golang.org/x/xerrors"
+
+	"github.com/coder/coder/v2/coderd/database"
+)
+
+type EnterpriseClaimer struct{}
+
+func (_ EnterpriseClaimer) Claim(
+	ctx context.Context,
+	store database.Store,
+	userID uuid.UUID,
+	name string,
+	presetID uuid.UUID,
+) (*uuid.UUID, error) {
+	var prebuildID *uuid.UUID
+	err := store.InTx(func(db database.Store) error {
+		// TODO: do we need this?
+		//// Ensure no other replica can claim a prebuild for this user simultaneously.
+		// err := store.AcquireLock(ctx, database.GenLockID(fmt.Sprintf("prebuild-user-claim-%s", userID.String())))
+		// if err != nil {
+		//	return xerrors.Errorf("acquire claim lock for user %q: %w", userID.String(), err)
+		//}
+
+		result, err := db.ClaimPrebuiltWorkspace(ctx, database.ClaimPrebuiltWorkspaceParams{
+			NewUserID: userID,
+			NewName:   name,
+			PresetID:  presetID,
+		})
+		if err != nil {
+			switch {
+			// No eligible prebuilds found
+			case errors.Is(err, sql.ErrNoRows):
+				// Exit, this will result in a nil prebuildID being returned, which is fine
+				return nil
```

Comment:
It will return `(nil, nil)`, I think generally this is discouraged?

Reply:
Responded in https://github.com/coder/coder/pull/17458/files#r2050948229